### ISSUE TITLE
Is it ok to use nngraph models inside of Sequencer?

### ISSUE THREAD

**Issue by yoonkim** [Element-Research/rnn#89]
Are there any issues with using nngraph inside a sequencer?

We've been trying to port over some code to elements, and we are seeing:
- large memory overhead (e.g. model that used to use 2GB memory is now using 12GB)
- # of parameters (and the results) seem to be slightly off

Thanks!

**Comment by kmnns** [comment]
+1 for that question. In general, is it OK to use *nngraph* with this package?

It seems to work, but I get funny graph plots and data structure printouts that I don't understand. For example, after putting your RNN-composed `nn.gModule` into `net`:

```lua
...

local net = nn.gModule({input}, {output})

for _,node in ipairs(net.forwardnodes) do
        print(node.data)
end
for _,node in ipairs(net.backwardnodes) do
        print(node.data)
end
```

This prints the exact same data structures (i.e. tables and subtables) for both *forwardnodes* and *backwardnodes*.

With common torch nn modules, *forwardnodes* and *backwardnodes* are clearly separated.
So far *nngraph* is just a black magic box for me. Does anyone have a better insight here?

**Comment by nicholas-leonard** [comment]
@yoonkim @kmnns Using rnn withing nngraph or nngraph within rnn are use cases I have not considered. I am not sure it would work. rnn extend abstract superclasses like nn.Module with methods like sharedClone, stepClone, backwardOnline, forget, etc. We also overwrite the type() method. nngraph also has its own extensions of the nn package, but I am not sure how they implement this. So there is definitely room for conflict between both packages. For example, I wouldn't be surprised that somehow sharedClone() doesn't work as expected with nngraph modules (that might explain 12GB leak). What we would need to resolve these, is some unit tests.

**Comment by kmnns** [comment]
Thanks @nicholas-leonard ! I wanted to setup a RNN-within-nngraph unit test anyway. 

I guess you would want a couple of `assertTensorEq` on the params and gradParams gradients spat out by two equivalent models with *nn* and *nngraph*, as in the current unit tests?

FYI, this is what is plotted when I naively stack a Sequencer on top of a BiSequencer in *nngraph*:
![screen shot 2015-12-23 at 10 27 57](https://cloud.githubusercontent.com/assets/5428444/11974067/c57b8520-a960-11e5-8964-111bbcb08287.png)

**Comment by nicholas-leonard** [comment]
@kmnns Yup, a couple of assertTensorEq should do.

**Comment by nicholas-leonard** [comment]
Hi, rnn should work nngraph's gModule with https://github.com/nicholas-leonard/dpnn/pull/32.

**Comment by blackyang** [comment]
Hi @nicholas-leonard  it seems that the link you provided above is broken now. Could you show me the link? Thanks!

**Comment by nicholas-leonard** [comment]
@blackyang Good point. Fixed it : https://github.com/Element-Research/dpnn/pull/32

**Comment by blackyang** [comment]
Thanks!

**Comment by blackyang** [comment]
Hi @nicholas-leonard is there any problem if I use a rnn module inside a gModule and manually do the fwd/bwd unrolling? 

e.g. 
```
local output = {}
for t =  1, targets_len do
    output[t] = some_gModule:forward({input1, input2, output[t-1})
end
```

Because from my understanding the above link only makes possible to use gModule inside a rnn. Correct me if I am wrong, many thanks!

**Comment by nicholas-leonard** [comment]
@blackyang There shouldn't be any issues with using an rnn within nngraph. Note however that the rnn handles the forwarding of `output[t-1]` internally.

**Comment by blackyang** [comment]
Hi @nicholas-leonard thanks for your clarification. For the `output[t-1]` I need it for other parts in my gModule :)
